flepimop/gempyor_pkg/src/gempyor/batch.py

```diff
@@ -1,23 +1,476 @@
 """
-Generate manifest files containing batch run metadata.
+Functionality for creating and submitting batch jobs.
 
-This module provides functionality for working with and executing batch jobs, either on
-HPC or AWS environments.
-
-Functions:
-    write_manifest: A utility to generate a manifest json file with metadata for a batch
-        job.
+This module provides functionality for required for batch jobs, including:
+* Objects for representing job resources and job sizes,
+* A system for registering and getting batch systems, and
+* Functionalities for writing job metadata to a manifest file.
 """
 
-__all__ = ["write_manifest"]
+__all__ = (
+    "BatchSystem",
+    "JobResources",
+    "JobSize",
+    "LocalBatchSystem",
+    "SlurmBatchSystem",
+    "get_batch_system",
+    "register_batch_system",
+    "write_manifest",
+)
 
 
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from datetime import timedelta
+from enum import Enum, auto
 import json
+import math
 from pathlib import Path
 import sys
-from typing import Any
+from typing import Any, Literal, overload
+import warnings
+
+from .utils import _git_head
+
+
+if sys.version_info >= (3, 11):
+    from typing import Self
+else:
+    Self = Any
+
+
+_batch_systems = []
+
+
+@dataclass(frozen=True, slots=True)
+class JobResources:
+    """
+    A batch submission job resources request.
+
+    Attributes:
+        nodes: The number of nodes to request.
+        cpus: The number of CPUs to request per a node.
+        memory: The amount of memory to request per a node in MB.
+
+    Raises:
+        ValueError: If any of the attributes are less than 1.
+
+    Examples:
+        >>> from gempyor.batch import JobResources
+        >>> resources = JobResources(nodes=5, cpus=10, memory=10*1024)
+        >>> resources
+        JobResources(nodes=5, cpus=10, memory=10240)
+        >>> resources.total_cpus
+        50
+        >>> resources.total_memory
+        51200
+        >>> resources.total_resources()
+        (5, 50, 51200)
+        >>> try:
+        ...     JobResources(nodes=0, cpus=1, memory=1024)
+        ... except Exception as e:
+        ...     print(e)
+        The 'nodes' attribute must be greater than 0, but instead was given '0'.
+    """
+
+    nodes: int
+    cpus: int
+    memory: int
+
+    def __post_init__(self) -> None:
+        for p in self.__slots__:
+            if (val := getattr(self, p)) < 1:
+                raise ValueError(
+                    (
+                        f"The '{p}' attribute must be greater than 0, "
+                        f"but instead was given '{val}'."
+                    )
+                )
```

Comment:
Per discussion - this might be simplified and more robust as a pydantic basemodel w/ positiveints as the attribute.

Reply:
Resolved in 6033c38